src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -142,4 +142,69 @@ public void close() throws IOException {
         ch.close();
     }
 
+    private static final int TRANSFER_SIZE = 8192;
+
+    @Override
+    public long transferTo(final OutputStream out) throws IOException {
+        if (out instanceof Channels.ChannelOutputStream) {
+            final var oc = ((Channels.ChannelOutputStream) out).channel();
+
+            if (ch instanceof FileChannel) {
+                final var fc = (FileChannel) ch;
+                final var pos = fc.position();
+                final var size = fc.size();
+                var i = 0L;
+                for (final var n = size - pos; i < n;
+                  i += fc.transferTo(pos + i, Long.MAX_VALUE, oc));
+                fc.position(size);
+                return i;
+            }
+
+            if (oc instanceof FileChannel) {
+                final var fc = (FileChannel) oc;
+                final var fcpos = fc.position();
+
+                if (ch instanceof SeekableByteChannel) {
+                    final var pos = ((SeekableByteChannel) ch).position();
+                    final var size = ((SeekableByteChannel) ch).size();
+                    var i = 0L;
+                    for (final var n = size - pos; i < n;
+                      i += fc.transferFrom(ch, fcpos + i, Long.MAX_VALUE));
+                    fc.position(fcpos + i);
+                    return i;
+                }
+
+                final var bb = ByteBuffer.allocateDirect(TRANSFER_SIZE);
```

Comment:
This will probably need to be changed to the use the TL buffer cache.

Reply:
Uhm... Maybe this is a dumb beginner's question, but: What is "the TL buffer cache"?

Reply:
I think I found what you mean and will use it: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/sun/nio/ch/Util.java#L221

Reply:
Done in https://github.com/openjdk/jdk/pull/4263/commits/8f589867154f8bc22ca6f6bb7a26d7ae841fe714. Thank you!